core/src/main/java/com/sequenceiq/cloudbreak/service/image/userdata/CcmUserDataService.java

```diff
@@ -0,0 +1,105 @@
+import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmConnectivityParameters;
+import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmParameterSupplier;
+import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmParameters;
+import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmV2ParameterSupplier;
+import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmV2Parameters;
+import com.sequenceiq.cloudbreak.ccm.endpoint.KnownServiceIdentifier;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceFamilies;
+import com.sequenceiq.cloudbreak.ccm.key.CcmResourceUtil;
+import com.sequenceiq.cloudbreak.common.service.HostDiscoveryService;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.instance.InstanceGroup;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+
+@Service
+public class CcmUserDataService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CcmUserDataService.class);
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private HostDiscoveryService hostDiscoveryService;
+
+    @Autowired
+    private CcmParameterSupplier ccmParameterSupplier;
+
+    @Autowired
+    private CcmV2ParameterSupplier ccmV2ParameterSupplier;
+
+    public CcmConnectivityParameters fetchAndSaveCcmParameters(Stack stack) {
+        CcmConnectivityParameters ccmConnectivityParameters = new CcmConnectivityParameters();
+        if (stack.getTunnel().useCcmV1()) {
+            String userCrn = ThreadBasedUserCrnProvider.getUserCrn();
+            String actorCrn = Objects.requireNonNull(userCrn, "userCrn is null");
+
+            ImmutableMap.Builder<KnownServiceIdentifier, Integer> builder = ImmutableMap.builder();
+            int gatewayPort = Optional.ofNullable(stack.getGatewayPort()).orElse(ServiceFamilies.GATEWAY.getDefaultPort());
+            builder.put(KnownServiceIdentifier.GATEWAY, gatewayPort);
+
+            // Optionally configure a tunnel for (nginx fronting) Knox
+            if (stack.getCluster().getGateway() != null) {
+                // JSA TODO Do we support a non-default port for the nginx that fronts Knox?
+                builder.put(KnownServiceIdentifier.KNOX, ServiceFamilies.KNOX.getDefaultPort());
+            }
+
+            Map<KnownServiceIdentifier, Integer> tunneledServicePorts = builder.build();
+            CcmParameters ccmV1Parameters = ccmParameterSupplier.getCcmParameters(actorCrn, ThreadBasedUserCrnProvider.getAccountId(),
+                    CcmResourceUtil.getKeyId(stack.getResourceCrn()), tunneledServicePorts).orElse(null);
+            ccmConnectivityParameters = new CcmConnectivityParameters(ccmV1Parameters);
+
+            saveCcmV1Config(stack.getId(), ccmV1Parameters);
+        } else if (stack.getTunnel().useCcmV2()) {
+            String gatewayHostName = hostDiscoveryService.generateHostname(stack.getCustomHostname(),
+                    stack.getGatewayHostGroup().map(InstanceGroup::getGroupName).orElse(""), 0L,
```

Comment:
- when could this happen: `.orElse("")`? shouldn't we fail instead?
- I'm not sure how HA for SDX works, maybe with a single gateway, but would that `0L` cause any problem?

Reply:
all the gateway nodes share the same mTLS cert so 0L would not cause any problem because mTLS doesnt validate client fqdn. 

orElse("")? would never happen because there are validations that atleast one hostGroup should be gatewayHostGroup.